### PR TITLE
Distinct storage device support

### PR DESCRIPTION
**Pull request by furkansahin**
This commit introduces a more intelligent allocator. Now, a host may
have multiple storage devices (not RAID0), and we can allocate a VM's
disks from each of those storage devices distinctively. Most of the
parts in this commit comes from
https://github.com/ubicloud/ubicloud/commit/9cf65be16ea3e9b632fab16503f5aa92c39a2b53 while changing the main
allocation query to handle an edge case. For example, the previous
commit does simply check for the the vm_storage_device with the highest
size and checks if there is a host that has x numbers of disk that can
host the vm_storage_device. This is not very accurate since a VM might
have 2 disks such as;
1. disk1: 30GB
2. disk2: 2TB

The previous commit would check if there is a host with 2 disks, each
has at least 2 TB capacity. With this commit, we try to check storage
devices per host and see if there is a mapping of vm_storage_devices to
storage_devices. The new commit would be able to allocate if there is a
host with, for example;
1. storage_device1.available_storage_gib: 100GB
2. storage_device2.available_storage_gib: 3TB

The previous commit would fail to do so.

Now, in this commit, there is something we need to address, what good
does it have to place vm disks in distinct storage devices? This is
mainly implemented for MinIO, so that it can survive, and self heal in
case a specific drive fails. There is also the question of, what about
the OS disk? Previously, I have considered putting OS disk of a VM into
the hosts OS disk. This is a good idea because, in any case, if the host
OS disk fails, all of the VMs in that host will already fail. The data
disks, on the other hand, will be just fine. However, making the
distinction of OS disk for VM, and allocating storage from other drives
will make this commit much more complicated, especially considering we
will have to dive into allocator in short term, anyway. Here, in this
PR, I made the decision to leave the placement of VM's OS disk or data
disk to fate. OS disk will be seen as any other disk, and we will
(probably) randomly choose any disk (e.g. host's OS disk, or any other
data disk) to place them in according to the available space. This way,
if one of the data disks fail, it is possible that some of the VMs in
that host will be unavailable, and some will lose the data disks.
However, for now, this is not an issue we need to tackle, because there
is exactly 1 minIO VM running on a Minio host. In addition, our
configuration of the host will have a smaller OS disk size and larger
data disk sizes. At the allocation phase, we can enforce the OS disk of
a VM to be placed in the host's os disk. To make it more clear, let's
assume this is our host configuration;
1. 1TB NVMe OS disk
2. 2x4TB NVMe extra disks

When, I provision a minio vm on that host, I will ask this to have;
1. 30GB of OS disk
2. 2x4TB of data disks.

In this configuration, allocator will be forced to place the OS disk
into the host's OS disk space.

### CODE REVIEW

**Comment by fdr** [comment]
I see you got impatient with me not handing off things for real and just started doing it :) sorry about that. How are you accounting for the case of multiplexing, which is what stymied me in the original patch? (or getting around it, by carefully putting the OS disk on the same disk as the host OS) with adequate control? TBD?

**Comment by furkansahin** [comment]
I will separate the Minio changes into its own PR.

**Review by fdr — APPROVED** [review]
OS device situation is a bit weird as you wrote, but if you have a handle on it, let's go ahead, I think that complicating the allocater (for this and other reasons) will happen sometime, but we can learn something by getting the patch as-is into production.

**Review by fdr — APPROVED** [review]
now that it's under your supervision, I'm okay with some of the weak spots of this patch (arbitrary OS disk assignment). Previously, when I tasked it to myself, knowing there could be no future close administration, I set the bar higher, but the equilibrium of that was...it was not going to get done in a reasonable amount of time.

File: prog/vm/nexus.rb [hunk 1]
```diff
@@ -494,14 +545,15 @@ def before_run
         nic.incr_destroy
       end
 
+      vm.vm_storage_volumes.each do |vol|
+        dev = vol.storage_device
+        dev ||= StorageDevice.dataset.where(vm_host_id: vm.vm_host_id, name: "DEFAULT").first
```
Comment (furkansahin) [inline prog/vm/nexus.rb:550]: this line is necessary until all of the VMs are refreshed or they all have the vm_storage_volume <-> storage_device relationship properly setup. I need it for runners, for the long running ones, I will manually set them and then remove this line in an additional PR.

**Pull request merged**